File: codeforces/1167/F.java

// Problem : F. Scalar Queries
// Contest : Codeforces - Educational Codeforces Round 65 (Rated for Div. 2)
// URL : https://codeforces.com/contest/1167/problem/F
// Memory Limit : 256 MB
// Time Limit : 2000 ms
// Powered by CP Editor (https://github.com/cpeditor/cpeditor)

import java.io.*;
import java.util.*;
import java.util.stream.*;

public class a implements Runnable{
	
    public static void main(String[] args) {
        new Thread(null, new a(), "process", 1<<26).start();
    }
	public void run() {
		FastReader scan = new FastReader();
        PrintWriter out = new PrintWriter(new OutputStreamWriter(System.out));
		//PrintWriter out = new PrintWriter("file.out");
		Task solver = new Task();
		//int t = scan.nextInt();
		int t = 1;
		for(int i = 1; i <= t; i++) solver.solve(i, scan, out);
		out.close();
	}

	static class Task {
		static final int inf = Integer.MAX_VALUE;

		public void solve(int testNumber, FastReader sc, PrintWriter pw) {
			//CHECK FOR QUICKSORT TLE
			//***********************//
			//CHECK FOR INT OVERFLOW
			//***********************//
			int n = sc.nextInt();
			tup[] arr = new tup[n];
			for(int i = 0; i < n; i++) {
				arr[i] = new tup(sc.nextInt(), i);
			}
			Arrays.sort(arr);
			bit left = new bit(n);
			bit righ = new bit(n);
			long sum = 0;
			for(int i = 0; i < n; i++) {
				sum += 1l * (arr[i].b + 1) * (n - arr[i].b) % 1000000007 * arr[i].a;
				sum %= 1000000007;
				sum += 1l * left.sum(arr[i].b - 1) % 1000000007 * (n - arr[i].b) % 1000000007 * arr[i].a;
				sum %= 1000000007;
				sum += 1l * righ.sum(n - 1, arr[i].b + 1) % 1000000007 * (arr[i].b + 1) % 1000000007 * arr[i].a;
				sum %= 1000000007;
				left.add(arr[i].b, arr[i].b + 1);
				righ.add(arr[i].b, n - arr[i].b);
			}
			pw.println(sum);
		}
		static class bit { 
			int n;
			long[] bit;
			public bit(int n) {
				this.n=n;
				bit=new long[n+1];
			}
			void add(int ind, long c) {
				for(; ind<n;ind|=(ind+1)) {
					bit[ind]+=c;
				}
			}
			long sum(int r) {
				long out =0;
				for(;r>=0;r=(r&(r+1))-1) {
					out+=bit[r];
				}
				return out;
			}
			long sum(int r, int l) {
				return sum(r)-sum(l-1);
			}
		}
	}
	static long binpow(long a, long b, long m) {
		a %= m;
		long res = 1;
		while (b > 0) {
			if ((b & 1) == 1)
				res = res * a % m;
			a = a * a % m;
			b >>= 1;
		}
		return res;
	}
	static void sort(int[] x){
		shuffle(x);
		Arrays.sort(x);
	}
	static void sort(long[] x){
		shuffle(x);
		Arrays.sort(x);
	}
	static class tup implements Comparable<tup>{
		int a, b;
		tup(int a,int b){
			this.a=a;
			this.b=b;
		}
		@Override
		public int compareTo(tup o){
			return Integer.compare(a,o.a);
		}
	}
	static void shuffle(int[] a) {
		Random get = new Random();
		for (int i = 0; i < a.length; i++) {
			int r = get.nextInt(i + 1);
			int temp = a[i];
			a[i] = a[r];
			a[r] = temp;
		}
	}

	static void shuffle(long[] a) {
		Random get = new Random();
		for (int i = 0; i < a.length; i++) {
			int r = get.nextInt(i + 1);
			long temp = a[i];
			a[i] = a[r];
			a[r] = temp;
		}
	}

	static class FastReader {
		BufferedReader br;
		StringTokenizer st;

		public FastReader() {
			br = new BufferedReader(new InputStreamReader(System.in));
		}

		public FastReader(String s) throws FileNotFoundException {
			br = new BufferedReader(new FileReader(new File(s)));
		}

		String next() {
			while (st == null || !st.hasMoreElements()) {
				try {
					st = new StringTokenizer(br.readLine());
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
			return st.nextToken();
		}

		int nextInt() {
			return Integer.parseInt(next());
		}

		long nextLong() {
			return Long.parseLong(next());
		}

		double nextDouble() {
			return Double.parseDouble(next());
		}

		String nextLine() {
			String str = "";
			try {
				str = br.readLine();
			} catch (IOException e) {
				e.printStackTrace();
			}
			return str;
		}
	}

}

File: codeforces/817/F.cpp

// Problem: F. MEX Queries
// Contest: Codeforces - Educational Codeforces Round 23
// URL: https://codeforces.com/contest/817/problem/F
// Memory Limit: 256 MB
// Time Limit: 2000 ms
// Powered by CP Editor (https://github.com/cpeditor/cpeditor)

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const int mx = 4e5 + 3;

int segt[mx * 4][2];
int lazy[mx * 8];
int lazyA[mx * 8];

void push(int v) {
  if (lazyA[v] == 1) {
    lazyA[2 * v] = 1;
    lazyA[2 * v + 1] = 1;
    segt[v][1] = segt[v][0] + segt[v][1];
    segt[v][0] = 0;
  }
  if (lazyA[v] == 2) {
    lazyA[2 * v] = 2;
    lazyA[2 * v + 1] = 2;
    segt[v][0] = segt[v][0] + segt[v][1];
    segt[v][1] = 0;
  }
  if (lazy[v] == -1) {
    lazy[2 * v] = -1;
    lazy[2 * v + 1] = -1;
  }
  if (lazy[v] == 1) {
    if (2 * v < mx * 2 && lazy[2 * v] == -1) {
      push(2 * v);
    }
    lazy[2 * v] ^= 1;
    if (2 * v + 1 < mx * 2 && lazy[2 * v + 1] == -1) {
      push(2 * v + 1);
    }
    lazy[2 * v + 1] ^= 1;
    swap(segt[v][0], segt[v][1]);
  }
  lazy[v] = 0;
  lazyA[v] = 0;
}

void init(int v, int tl, int tr) {
  if (tl == tr) {
    segt[v][0] = 1;
  } else {
    int mid = (tl + tr) >> 1;
    init(v * 2, tl, mid);
    init(v * 2 + 1, mid + 1, tr);
    segt[v][0] = segt[2 * v][0] + segt[2 * v + 1][0];
  }
}

int sum(int v, int tl, int tr, ll l, ll r) {
  push(v);
  if (l > tr || r < tl)
    return 0;
  else if (tl >= l && tr <= r) {
    return segt[v][1];
  } else {
    int mid = (tl + tr) >> 1;
    return sum(v * 2, tl, mid, l, r) + sum(v * 2 + 1, mid + 1, tr, l, r);
  }
}

void update1(int v, int tl, int tr, ll l, ll r) {
  push(v);
  if (l > tr || r < tl)
    return;
  else if (tl >= l && tr <= r) {
    lazyA[v] = 1;
    lazy[v] = -1;
    push(v);
  } else {
    int mid = (tl + tr) >> 1;
    update1(v * 2, tl, mid, l, r);
    update1(v * 2 + 1, mid + 1, tr, l, r);
    segt[v][1] = segt[2 * v][1] + segt[2 * v + 1][1];
    segt[v][0] = segt[2 * v][0] + segt[2 * v + 1][0];
  }
}

void update2(int v, int tl, int tr, ll l, ll r) {
  push(v);
  if (l > tr || r < tl)
    return;
  else if (tl >= l && tr <= r) {
    lazyA[v] = 2;
    lazy[v] = -1;
    push(v);
  } else {
    push(v);
    int mid = (tl + tr) >> 1;
    update2(v * 2, tl, mid, l, r);
    update2(v * 2 + 1, mid + 1, tr, l, r);
    segt[v][1] = segt[2 * v][1] + segt[2 * v + 1][1];
    segt[v][0] = segt[2 * v][0] + segt[2 * v + 1][0];
  }
}

void update3(int v, int tl, int tr, ll l, ll r) {
  push(v);
  if (l > tr || r < tl)
    return;
  else if (tl >= l && tr <= r) {
    push(v);
    lazy[v] ^= 1;
    push(v);
  } else {
    int mid = (tl + tr) >> 1;
    update3(v * 2, tl, mid, l, r);
    update3(v * 2 + 1, mid + 1, tr, l, r);
    segt[v][1] = segt[2 * v][1] + segt[2 * v + 1][1];
    segt[v][0] = segt[2 * v][0] + segt[2 * v + 1][0];
  }
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  init(1, 0, mx - 1);
  set<ll> v = {};
  v.insert(1);
  int q;
  cin >> q;
  ll qs[q][3];
  for (int i = 0; i < q; i++) {
    cin >> qs[i][0] >> qs[i][1] >> qs[i][2];
    v.insert(qs[i][1]);
    v.insert(qs[i][2]);
    v.insert(qs[i][2] + 1);
  }
  map<ll, int> mp = {};
  int k = 0;
  auto x = v.begin();
  vector<ll> vd = {};
  for (; x != v.end(); x++, k++) {
    mp.insert(make_pair(*x, k));
    vd.push_back(*x);
  }
  for (int i = 0; i < q; i++) {
    qs[i][1] = mp.find(qs[i][1])->second;
    qs[i][2] = mp.find(qs[i][2])->second;
  }
  for (int i = 0; i < q; i++) {
    if (qs[i][0] == 1) {
      update1(1, 0, mx - 1, qs[i][1], qs[i][2]);
    }
    if (qs[i][0] == 2) {
      update2(1, 0, mx - 1, qs[i][1], qs[i][2]);
    }
    if (qs[i][0] == 3) {
      update3(1, 0, mx - 1, qs[i][1], qs[i][2]);
    }
    if (sum(1, 0, mx - 1, 0, 0) == 0) {
      cout << 1 << endl;
    } else {
      int l = 0;
      int r = mx - 1;
      while (r - l > 1) {
        int mid = (l + r) >> 1;
        if (sum(1, 0, mx - 1, 0, mid) == mid + 1) {
          l = mid;
        } else {
          r = mid - 1;
        }
      }
      if (sum(1, 0, mx - 1, 0, r) == r + 1) {
        l = r;
      }
      cout << vd[l + 1] << endl;
    }
  }
}